cms/djangoapps/contentstore/views/utilities/tasks.py

```diff
@@ -0,0 +1,97 @@
+from celery.task import task
+from django.conf import settings
+from django.core.mail import send_mail
+from django.core.urlresolvers import reverse
+from smtplib import SMTPException
+
+from edxmako.shortcuts import render_to_string
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.inheritance import own_metadata
+
+from opaque_keys.edx.keys import CourseKey
+from xblock.fields import Scope
+
+
+def _get_course_problems(course_key):
+    """
+    Retrieve all problems in the course
+    """
+    # Note: Some components may have been orphaned due to a bug in split
+    # modulestore (PLAT-799). We may be returning orphaned components in this
+    # list to be updated as well
+    return modulestore().get_items(
+        course_key,
+        qualifiers={"category": 'problem'},
+    )
+
+
+def _send_email_on_completion(course, user_username, user_email, modified_settings, success):
+    """
+    Send email to user on completion of celery task completion
+    """
+    context = {
+        'user_username': user_username,
+        'course_name': course.display_name,
+        'modified_settings': modified_settings,
+        'success': success
+    }
+
+    from_email = settings.DEFAULT_FROM_EMAIL
+    subject = render_to_string('emails/utilities_bulk_update_done_subject.txt', context)
+    subject = ''.join(subject.splitlines())
+    message = render_to_string('emails/utilities_bulk_update_done_message.txt', context)
+
+    try:
+        send_mail(
+            subject,
+            message,
+            from_email,
+            [user_email],
+            fail_silently=False
+        )
+    except SMTPException:
+        print("Failure sending e-mail for bulk update completion to %s", user_email)
+
+
+class BulkUpdateUtil():
+    """
+    Utility class that hold functions for bulksettings operations
+    """
+
+    @classmethod
+    def update_metadata(cls, course_key, user_id, metadata):
+        """
+        Updates metadata settings for all problems.
+        """
+
+        problems = _get_course_problems(course_key)
+
+        with modulestore().bulk_operations(course_key):
+            for problem in problems:
+                for metadata_key, value in metadata.items():
+                    field = problem.fields[metadata_key]
+                    try:
+                        value = field.from_json(value)
+                    except Exception as exception:
+                        print(exception)
+                    field.write_to(problem, value)
+
+                modulestore().update_item(problem, user_id)
+
+                if modulestore().has_published_version(problem):
```

Comment:
Do we need/want to request the singleton via `modulestore()` each time?
Can we memoize it, ala `store = modulestore(); if store.has_published_version(...)`?

Especially in a loop that could be called hundreds/thousands of times,
those extra calls (inside `modulestore()`') could start to add up...

Reply:
no need, replaced with store